common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java

```diff
@@ -106,28 +127,32 @@ public MaterialConfig config() {
     }
 
     public List<Modification> modificationsSince(File baseDir, Revision revision, final SubprocessExecutionContext execCtx) {
+        unshallowIfNeeded(ProcessOutputStreamConsumer.inMemoryConsumer(), revision, baseDir);
         GitCommand gitCommand = getGit(baseDir);
         return gitCommand.modificationsSince(revision);
     }
 
     public MaterialInstance createMaterialInstance() {
-        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString());
+        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString(), shallowClone);
     }
 
     @Override
     protected void appendCriteria(Map<String, Object> parameters) {
         parameters.put(ScmMaterialConfig.URL, url.forCommandline());
         parameters.put("branch", branch);
+        parameters.put("shallowClone", shallowClone);
```

Comment:
>  Ignoring the shallowClone option on the server side and doing whatever is needed to make polling work (user's don't care about that clone. It's internal to the Go Server). On the agent side, do what the users asked for.

I like this. The trouble I have is: material is a not per pipeline model, even in cruise-config.xml it looks like. So it is currently not possible to have per pipeline attributes on materials except you persist separated material instances into DB.

This afternoon when I dig deeper into the code, I found ScheduledPipelineLoader#pipelineWithPasswordAwareBuildCauseByBuildId method is very interesting.
```java
    public Pipeline pipelineWithPasswordAwareBuildCauseByBuildId(final long buildId) {
          ...
            if (materialConfig instanceof PasswordAwareMaterial) {
                PasswordAwareMaterial configuredMaterial = (PasswordAwareMaterial) materialConfig;
                ((PasswordAwareMaterial) usedMaterial).setPassword(configuredMaterial.getPassword());
            }
       ...
   }
```
How about we load material config from pipeline config here and merge settings into the material?
```java
    public Pipeline pipelineWithPasswordAwareBuildCauseByBuildId(final long buildId) {
         ...
           usedMaterial.updateFrom(materialConfigFromPipelineConfig)
        ...
   }
```
This way we do not even need write shallowClone flag into the database.